### ISSUE TITLE
Issues with specific values

### ISSUE THREAD

**Issue by Jabe** [dotMorten/Microsoft.SqlServer.Types#35]
There is an issue with specific values. Using the following code in linqpad:

```c#
var parsed = HierarchyId.Parse("/1.3.2/").Dump("parsed");
var ms = new MemoryStream();
parsed.Write(new BinaryWriter(ms));
ms.Position = 0;
var roundTrip = new Microsoft.SqlServer.Types.SqlHierarchyId();
roundTrip.Read(new BinaryReader(ms));
roundTrip.Dump("roundTrip");
```

|Input|Value of parsed|Value of roundTrip|
|-----|-----|-----|
|/1.0.2/|/1.0.2/|/1.0.2/|
|/1.1.2/|/1.1.2/|/1.1.2/|
|/1.2.2/|/1.2.2/|/1.2.2/|
|/1.3.2/|/1.3.2/|InvalidCastException: No pattern found for: 11010|
|/3.0/|/3.0/|/|

Using `Microsoft.SqlServer.Types.SqlHierarchyId (Microsoft.SqlServer.Types, Version=1.1.0.0, Culture=neutral, PublicKeyToken=null)` directly in netcore31 (linqpad6) produces the same bad result.

Using `Microsoft.SqlServer.Types.SqlHierarchyId (Microsoft.SqlServer.Types, Version=13.0.0.0, Culture=neutral, PublicKeyToken=89845dcd8080cc91)` directly in net46+ (linqpad5) works with all tested values as expected.

(via https://github.com/efcore/EFCore.SqlServer.HierarchyId/issues/7)

**Comment by ppasieka** [comment]
@dotMorten do you have any materials where I could learn more about the bit patterns used in implementing HierarchyId?

Or maybe you have some ideas how to fix this issue?

**Comment by dotMorten** [comment]
> do you have any materials where I could learn more about the bit patterns used in implementing HierarchyId?

Nope. I didn't even provide this implementation (community contribution), and honestly don't even know what it is used for.

**Comment by ppasieka** [comment]
@olmobrutall Would you be able to help with this issue? Or provide some materials where I could learn more about these bits patterns?
Definitely, serialization part is broken. When I try to look at data in SQL management studio I'm getting an exception
after using `.ToString` on HierarchyId
```
A .NET Framework error occurred during execution of user-defined routine or aggregate "hierarchyid": 
Microsoft.SqlServer.Types.HierarchyIdException: 24000: SqlHierarchyId operation failed because HierarchyId object was constructed from an invalid binary string. 
Microsoft.SqlServer.Types.HierarchyIdException: 
   at Microsoft.SqlServer.Types.ex_raise(Int32 major, Int32 minor, Int32 sev, Int32 state, Object param1, Object param2, Object param3)
   at Microsoft.SqlServer.Types.OrdPath.GetBits(UInt16 startBit, UInt16 nBits, UInt64& bits)
   at Microsoft.SqlServer.Types.OrdPath.ExtractOrd(UInt16& bitOffset, UInt32 stage, Int64& ord)
   at Microsoft.SqlServer.Types.OrdPath.ExtractComponent(UInt16& bitOffset, SComponent& component, levelType& type)
   at Microsoft.SqlServer.Types.OrdPath.ToString()
```

**Comment by olmobrutall** [comment]
Hi @ppasieka, 

the bit patterns come from http://www.adammil.net/blog/v100_how_the_SQL_Server_hierarchyid_data_type_works_kind_of_.html

Not sure if I’ll find time to work on it, but I can review a potential change if you go for it. 

Cheers

**Comment by olmobrutall** [comment]
It's definitely an issue with serialization, I've converted the code to a Unit Test and also compare with values comming from the database: 

```C#
        [DataTestMethod]
        [DataRow("/1.0.2/")]
        [DataRow("/1.1.2/")]
        [DataRow("/1.2.2/")]
        [DataRow("/1.3.2/")]
        [DataRow("/3.0/")]
        public void SerializeDeserialize(string route)
        {
            var parsed = SqlHierarchyId.Parse(route);
            //var ms = new MemoryStream();
            //parsed.Write(new BinaryWriter(ms));
            //ms.Position = 0;
            //var roundTrip = new Microsoft.SqlServer.Types.SqlHierarchyId();
            //roundTrip.Read(new BinaryReader(ms));
            //Assert.AreEqual(parsed, roundTrip);

            using (SqlConnection con = new SqlConnection(@"Data Source=.\SQLEXPRESS;Initial Catalog=HierarchyTest;Integrated Security=true"))
            {
                con.Open();
                var id = new SqlCommand($"INSERT INTO [dbo].[TreeNode] (Route) output INSERTED.ID VALUES ('{route}') ", con).ExecuteScalar();

                using (var reader = new SqlCommand($"SELECT Route FROM [dbo].[TreeNode] WHERE ID = " + id, con).ExecuteReader())
                {
                    while (reader.Read())
                    {
                        var roundTrip = new Microsoft.SqlServer.Types.SqlHierarchyId();
                        roundTrip.Read(new BinaryReader(reader.GetStream(0)));
                        Assert.AreEqual(parsed, roundTrip);
                    }
                }
            }
```

The deserializer works, the serializer produces the error. Maybe tomorrow I can take a deeper look.

**Comment by ppasieka** [comment]
@olmobrutall Thanks. I will also look into it tomorrow :) And yes I came to the same conclusion that the issue is with a serializer. Deserialization part work for me without any problems so far

**Comment by olmobrutall** [comment]
Fixed in https://github.com/dotMorten/Microsoft.SqlServer.Types/pull/55

**Comment by ppasieka** [comment]
@olmobrutall amazing 👍 I will check it against my project. Thanks for that fast response 😊

**Comment by ppasieka** [comment]
@dotMorten Fix provided in #55 works nicely. What do you think about creating a release? With PRs #55 and #47 out, the list of HierarychId problems would be reduced to only one (#36)
